docs/index.bs

```diff
@@ -3154,7 +3150,30 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
                       1. If |client| is not null, |response|'s [=response/type=] is "`opaque`", and [=cross-origin resource policy check=] with |request|'s [=request/origin=], |client|, "", and |response|'s [=filtered response/internal response=] returns <b>blocked</b>, then return null.
                       1. Return |response|.
               1. Return null.
-      1. If |request| is a <a>non-subresource request</a>, |request| is a [=navigation request=], |registration|'s [=navigation preload enabled flag=] is set, |request|'s [=request/method=] is \`<code>GET</code>\`, |registration|'s [=active worker=]'s [=set of event types to handle=] [=set/contains=] <code>fetch</code>, and |registration|'s [=active worker=]'s [=all fetch listeners are empty flag=] is not set then:
+          1. Else if |source| is {{RouterSourceEnum/"race-network-and-fetch-handler"}}, and |request|'s [=request/method=] is \`<code>GET</code>\` then:
+
+              Note: In order to avoid duplicated fetches to |request|, the user agent may reuse |raceNetworkRequestResponse| as the result of {{FetchEvent}}'s [=FetchEvent/potential response=], or the fallback request.
+
+              1. Let |queue| be an empty [=queue=] of [=/response=].
+              1. Let |raceFetchController| be null.
+              1. Run the following substeps [=in parallel=], but [=abort when=] |controller|'s [=fetch controller/state=] is "<code>terminated</code>" or "<code>aborted</code>":
+                  1. Set |raceFetchController| to the result of [=Fetch|fetching=] |request|.
+                  1. To [=fetch/processResponse=] for |raceNetworkRequestResponse|, run these substeps:
+                      1. If |raceNetworkRequestResponse|'s [=response/type=] is "`error`", then:
+                          1. Terminate these substeps.
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is `200`, then:
+                          1. [=queue/Enqueue=] |raceNetworkRequestResponse| to |queue|.
+              1. [=If aborted=], then:
+                  1. Let |deserializedError| be the result of [=deserialize a serialized abort reason=] given null and |workerRealm|.
+                  1. [=fetch controller/Abort=] |raceFetchController| with |deserializedError|.
+              1. Run the following substeps [=in parallel=]:
+                  1. Resolve |preloadResponse| with undefined.
+                  1. Let |fetchHandlerResponse| to the result of [=Create Fetch Event and Dispatch=] with |request|, |registration|, |useHighResPerformanceTimers|, |timingInfo|, |workerRealm|, |reservedClient|, and |preloadResponse|.
+                  1. [=queue/Enqueue=] |fetchHandlerResponse| to |queue|.
+              1. Wait until |queue| is not empty.
+              1. Set |response| to the result of [=dequeue=] |queue|.
+              1. Return |response|.
```

Comment:
Just before this step, don't we abort |raceFetchController| if it is still running?

Reply:
We may want to reuse the response from the network request as a fallback, abort it only when |fetchHandlerResponse| is a successful response. This aligns with the current chromium implementation.
https://source.chromium.org/chromium/chromium/src/+/main:content/browser/service_worker/service_worker_main_resource_loader.cc;l=722;drc=7dc3003765c74daa4bb49e4463e3a7963d8c7ea1;bpv=1;bpt=1?q=main_resource_loader&ss=chromium%2Fchromium%2Fsrc